File: vi/codes/csharp/chapter_backtracking/permutations_i.cs
/**
 * File: permutations_i.cs
 * Created Time: 2023-04-24
 * Author: hpstory (hpstory1024@163.com)
 */

namespace hello_algo.chapter_backtracking;

public class permutations_i {
    /* Giải thuật quay lui: Hoán vị I */
    void Backtrack(List<int> state, int[] choices, bool[] selected, List<List<int>> res) {
        // Khi độ dài trạng thái bằng số lượng phần tử, ghi lại lời giải
        if (state.Count == choices.Length) {
            res.Add(new List<int>(state));
            return;
        }
        // Duyệt qua tất cả các lựa chọn
        for (int i = 0; i < choices.Length; i++) {
            int choice = choices[i];
            // Cắt tỉa: không cho phép chọn lại phần tử đã chọn
            if (!selected[i]) {
                // Thử: thực hiện lựa chọn, cập nhật trạng thái
                selected[i] = true;
                state.Add(choice);
                // Tiến hành vòng lựa chọn tiếp theo
                Backtrack(state, choices, selected, res);
                // Quay lui: hoàn tác lựa chọn, khôi phục về trạng thái trước đó
                selected[i] = false;
                state.RemoveAt(state.Count - 1);
            }
        }
    }

    /* Hoán vị I */
    List<List<int>> PermutationsI(int[] nums) {
        List<List<int>> res = [];
        Backtrack([], nums, new bool[nums.Length], res);
        return res;
    }

    [Test]
    public void Test() {
        int[] nums = [1, 2, 3];

        List<List<int>> res = PermutationsI(nums);

        Console.WriteLine("Input array nums = " + string.Join(", ", nums));
        Console.WriteLine("All permutations res = ");
        foreach (List<int> permutation in res) {
            PrintUtil.PrintList(permutation);
        }
    }
}

File: vi/codes/swift/chapter_backtracking/permutations_i.swift
/**
 * File: permutations_i.swift
 * Created Time: 2023-04-30
 * Author: nuomi1 (nuomi1@qq.com)
 */

/* Giải thuật quay lui: Hoán vị I */
func backtrack(state: inout [Int], choices: [Int], selected: inout [Bool], res: inout [[Int]]) {
    // Khi độ dài trạng thái bằng số lượng phần tử, ghi lại lời giải
    if state.count == choices.count {
        res.append(state)
        return
    }
    // Duyệt tất cả các lựa chọn
    for (i, choice) in choices.enumerated() {
        // Cắt tỉa: không cho phép chọn lại phần tử đã dùng
        if !selected[i] {
            // Thử: thực hiện lựa chọn, cập nhật trạng thái
            selected[i] = true
            state.append(choice)
            // Tiến hành vòng lựa chọn tiếp theo
            backtrack(state: &state, choices: choices, selected: &selected, res: &res)
            // Quay lui: hoàn tác lựa chọn, khôi phục về trạng thái trước đó
            selected[i] = false
            state.removeLast()
        }
    }
}

/* Hoán vị I */
func permutationsI(nums: [Int]) -> [[Int]] {
    var state: [Int] = []
    var selected = Array(repeating: false, count: nums.count)
    var res: [[Int]] = []
    backtrack(state: &state, choices: nums, selected: &selected, res: &res)
    return res
}

@main
enum PermutationsI {
    /* Mã điều khiển */
    static func main() {
        let nums = [1, 2, 3]

        let res = permutationsI(nums: nums)

        print("Input array nums = \(nums)")
        print("All permutations res = \(res)")
    }
}
